File: ELEC881_Tabnine/src/No Description Prompt/Java script (JS)/q28-max-sum-of-rectangle-no-larger-than-k.js
/**
 * @param {number[][]} matrix
 * @param {number} k
 * @return {number}
 */
var maxSumSubmatrix = function(matrix, k) {
    let rows = matrix.length;
    let cols = matrix[0].length;

    let prefixSum = Array.from({ length: rows + 1 }, () => Array.from({ length: cols + 1 }, () => 0));

    for (let i = 1; i <= rows; i++) {
        for (let j = 1; j <= cols; j++) {
            prefixSum[i][j] = prefixSum[i - 1][j] + prefixSum[i][j - 1] - prefixSum[i - 1][j - 1] + matrix[i - 1][j - 1];
        }
    }

    let minHeap = [];
    let result = -Infinity;
    for (let left = 0; left < cols; left++) {
        for (let right = left; right < cols; right++) {
            while (minHeap.length && prefixSum[minHeap[0][0]][right] - prefixSum[minHeap[0][0]][left - 1] > k) {
                result = Math.max(result, prefixSum[minHeap.shift()[0]][right] - prefixSum[minHeap[0][0]][left - 1]);
            }

            for (let top = 0; top < rows; top++) {
                while (minHeap.length && prefixSum[top + 1][right] - prefixSum[top][left - 1] >= prefixSum[minHeap[minHeap.length - 1][0]][right]) {
                    minHeap.pop();
                }
                minHeap.push([top, prefixSum[top + 1][right] - prefixSum[top][left - 1]]);
            }
        }
    }
    return result;
};

File: ELEC881_Tabnine/src/No Description Prompt/Python/q28-max-sum-of-rectangle-no-larger-than-k.py
class Solution(object):
    def maxSumSubmatrix(self, matrix, k):
        """
        :type matrix: List[List[int]]
        :type k: int
        :rtype: int
        """
        if not matrix or not matrix[0]:
            return 0

        rows, cols = len(matrix), len(matrix[0])
        prefix_sum = [[0] * (cols + 1) for _ in range(rows + 1)]

        for i in range(1, rows + 1):
            for j in range(1, cols + 1):
                prefix_sum[i][j] = prefix_sum[i - 1][j] + prefix_sum[i][j - 1] - prefix_sum[i - 1][j - 1] + matrix[i - 1][j - 1]

        res = float('-inf')
        for left in range(1, cols + 1):
            for right in range(left, cols + 1):
                cur_sum = 0
                for top in range(1, rows + 1):
                    cur_sum += prefix_sum[top][right] - prefix_sum[top][left - 1]
                    for bottom in range(top, rows + 1):
                        cur_sum -= prefix_sum[bottom][right] + prefix_sum[bottom][left - 1]
                        if cur_sum >= k:
                            res = max(res, cur_sum)

        return res








































        